### ISSUE TITLE
Start using proper URL objects instead of string concatenation

### ISSUE THREAD

**Issue by chrisbobbe** [zulip/zulip-mobile#4081]
~~EDIT: Blocked on #3781, the RN 0.61 upgrade, unfortunately; see my comments here, up until https://github.com/zulip/zulip-mobile/issues/4081#issuecomment-622606771, below.~~ Looks like this is unblocked; see https://github.com/zulip/zulip-mobile/issues/4081#issuecomment-623716266 for a way to avoid depending on a file newly introduced in RN v0.61.

-----

Arising especially from https://github.com/zulip/zulip-mobile/pull/3819#issuecomment-576487586, https://github.com/zulip/zulip-mobile/pull/3819#issuecomment-578293246, and https://github.com/zulip/zulip-mobile/issues/3567#issuecomment-622575914, which all deal with something going wrong with constructing URLs, here's an issue for using something more reliable and consistent than string concatenation and various regexes.

Issues like https://github.com/zulip/zulip-mobile/issues/4058 may benefit from this too?

#4082 is a workaround that has our first TODO pointing to the eventual resolution of this issue.

It looks like React Native has a polyfill for the [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) Web API, but it's been cut down to reduce JS bundle size. https://github.com/charpeni/react-native-url-polyfill#why-do-we-need-this is a more complete polyfill (it says it would add 73.41 KB to the bundle), and it lists some bugs that people have run into with React Native core's polyfill.

**Comment by chrisbobbe** [comment]
Ha — in fact, a React Native maintainer closed an issue, at https://github.com/facebook/react-native/issues/16434#issuecomment-580094423, citing `react-native-url-polyfill` as (apparently) the official workaround.

**Comment by chrisbobbe** [comment]
I think https://github.com/facebook/react-native/issues/25717, and probably others, listed there, is enough to say that we shouldn't use React Native's polyfill.

Consider the bug in https://github.com/zulip/zulip-mobile/issues/3567#issuecomment-622574435:

```
webAuth.openBrowser(`${this.props.realm}/${url}`, otp);
```

incorrectly doubles a slash if `url` (which is a `.login_url` on an item in `.external_authentication_methods` from `/server_settings`) starts with a slash:

```
https://chat.zulip.org//accounts/login/social/google
```

All the ones I've seen do, but we shouldn't expect that they always will. If it's missing, then https://github.com/facebook/react-native/issues/25717 means we'll get

```
https://chat.zulip.orgaccounts/login/social/google
```

from

```
new URL(url, this.props.realm).toString()
```

.

**Comment by chrisbobbe** [comment]
Hmm, I don't think we can use any existing version of `react-native-url-polyfill`, though, until RN v0.61 (that's https://github.com/zulip/zulip-mobile/issues/3781).

[This file](https://github.com/charpeni/react-native-url-polyfill/blob/master/js/URL.js) in `react-native-url-polyfill`, which was [added](https://github.com/charpeni/react-native-url-polyfill/commit/37bb848a0c8ef60b29df12de825349f0824304d7#diff-375a2b7b7d929526f0dd58cd8bb8665b) in 1.0.0, [the initial release](https://github.com/charpeni/react-native-url-polyfill/releases/tag/v1.0.0), depends on [this file](https://github.com/facebook/react-native/blob/master/Libraries/Blob/NativeBlobModule.js) in React Native, which [was added](https://github.com/facebook/react-native/commit/342c81d75448e760ae32d62ebd9004bf6eeeff46) in v0.61.0.

On RN v0.59.10, which we're using, we get this error from Metro:

```
error: bundling failed: Error: Unable to resolve module `react-native/Libraries/Blob/NativeBlobModule` from `/Users/chrisbobbe/dev/zulip-mobile/node_modules/react-native-url-polyfill/js/URL.js`: Module `react-native/Libraries/Blob/NativeBlobModule` does not exist in the Haste module map
```

**Comment by chrisbobbe** [comment]
> Hmm, I don't think we can use any existing version of react-native-url-polyfill, though, until RN v0.61 (that's #3781).

Looks like the maintainer is quite responsive, though! https://github.com/charpeni/react-native-url-polyfill/issues/154

**Comment by ray-kraesig** [comment]
Rather than shipping a JavaScript library, would it make sense for us to make a small native plugin which just wraps an [NSURL](https://developer.apple.com/documentation/foundation/nsurl) (iOS) or [URL](https://developer.android.com/reference/java/net/URL) (Android) object?

**Comment by chrisbobbe** [comment]
> Rather than shipping a JavaScript library, would it make sense for us to make a small native plugin which just wraps an [NSURL](https://developer.apple.com/documentation/foundation/nsurl) (iOS) or [URL](https://developer.android.com/reference/java/net/URL) (Android) object?

Hmm, I wonder! Some quick Googling hasn't turned up anything off-the-shelf that does this. One possible explanation is because it's easy enough to do on one's own.

One question I'd have is how different those iOS and Android APIs are from each other, and whether we'd spend a lot of time reconciling differences between them, to give a single API.

Also, all of its methods would have to be asynchronous, right, because the React Native bridge is asynchronous. Might this make some race conditions hard to rule out (even if they're unlikely)?

**Comment by ray-kraesig** [comment]
> One question I'd have is whether we'd expect to spend a lot of time reconciling differences between the iOS and Android APIs.

I doubt it. The syntax and components of a URL are largely specified by RFC 3986. There are a fair number of idiosyncratic auxiliary functions on both of those classes, but we probably don't care about those.

> Also, all of its methods would have to be asynchronous, right, because the React Native bridge is asynchronous.

Oh right. I think that answers my previous question ("would it make sense?") with a solid "no" – race conditions or not. :sweat:

**Comment by gnprice** [comment]
> > Hmm, I don't think we can use any existing version of react-native-url-polyfill, though, until RN v0.61 (that's #3781).
> 
> Looks like the maintainer is quite responsive, though, which is a good general sign! [charpeni/react-native-url-polyfill#154 (comment)](https://github.com/charpeni/react-native-url-polyfill/issues/154#issuecomment-622611279)

Indeed!

At a quick look at the implementation, it looks like most of the work happens inside here:
https://github.com/charpeni/whatwg-url
(same maintainer's fork of another project). So we could probably just import that directly and call it explicitly -- it just won't be named `URL`.

(Or anyway it shouldn't be polyfilled as the global `URL`, because it won't have the `createObjectURL` implementation that that polyfill library adds -- and that I'm guessing RN's own implementation has, and RN relies on.)

**Comment by chrisbobbe** [comment]
Makes sense!

Something that `react-native-url-polyfill` (and the `whatwg-url-without-unicode` you mentioned, which does most of the work) depends on is assigning an export of [`buffer`](https://www.npmjs.com/package/buffer) to the global variable Buffer, where, previously (at least in our current master) nothing is assigned. React Native core's buggy [polyfill](https://github.com/facebook/react-native/blob/8c0c860e38f57e18296f689e47dfb4a54088c260/Libraries/Blob/URL.js#L115-L222) for URL has nothing to do with Buffer.

**Comment by gnprice** [comment]
> Something that `react-native-url-polyfill` (and the `whatwg-url-without-unicode` you mentioned, if we use it with React Native) depends on is assigning an export of [`buffer`](https://www.npmjs.com/package/buffer) to the global variable Buffer, where, previously (at least in our current master) nothing is assigned.

Hmm, where are you seeing that? I wouldn't have expected `whatwg-url-without-unicode` to try to do any mucking with globals.


> Also, we'd need to get a Flow libdef working for `react-native-url-polyfill`. Sadly, FlowTyped doesn't seem to have one,

Fortunately the API is quite small -- I think we only care about the `URL` class, which mostly looks like just a bag of properties. Here's the spec:
https://url.spec.whatwg.org/#url-class

**Comment by chrisbobbe** [comment]
Opened as https://github.com/charpeni/whatwg-url/pull/3.

Ah, I think I should have been more specific.

- `react-native-url-polyfill` mucks with globals, at https://github.com/charpeni/react-native-url-polyfill/blob/master/index.js#L18.
- Using `whatwg-url-without-unicode` directly would mean mucking with globals ourselves.

The reason for both is that at least one [file](https://github.com/charpeni/whatwg-url/blob/master/src/url-state-machine.js#L380) in `whatwg-url-without-unicode` depends on a global Buffer.

As discussed on the phone earlier today, I've opened https://github.com/charpeni/react-native-url-polyfill/blob/master/index.js#L18 to see if we can use `whatwg-url-without-unicode` directly, and not have to do anything with globals. (If that change is accepted, an eventual release of `react-native-url-polyfill`, which we'll likely use in a RN v0.61+ world, might adapt to the change, which would be nice as well.)

**Comment by chrisbobbe** [comment]
> > Also, we'd need to get a Flow libdef working for `react-native-url-polyfill`. Sadly, FlowTyped doesn't seem to have one,
> 
> Fortunately the API is quite small -- I think we only care about the `URL` class, which mostly looks like just a bag of properties. Here's the spec:
> https://url.spec.whatwg.org/#url-class

For anyone looking to make a PR, we have a new doc on putting together libdefs, where off-the-shelf ones don't exist! That's [here](https://github.com/zulip/zulip-mobile/blob/master/docs/howto/libdefs.md).

**Comment by chrisbobbe** [comment]
Linking recent conversation on CZO [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/URL.20library/near/878017).

**Comment by charpeni** [comment]
👋 Hey there.

I'm the maintainer of [`react-native-url-polyfill`](https://github.com/charpeni/react-native-url-polyfill) and [`whatwg-url-without-unicode`](https://github.com/charpeni/whatwg-url). Sorry for the delay, I've been chilling out in vacation, but I'm back now.

I plan to work on `react-native-url-polyfill` this week. Even though React Native 0.59 is an older version, I would like to make it compatible with a larger set of RN versions.

Also, thanks for looking into `buffer`, it seems fair to move that to `whatwg-url-without-unicode`, I'll just make sure there's no other dependencies in RN.

You will be hearing from me again later this week.

**Comment by chrisbobbe** [comment]
Hi, thanks for reaching out! 🙂 Let us know if and how we can be helpful.

**Comment by chrisbobbe** [comment]
Hi again, @charpeni! I wonder if you might be interested to follow some discussion we had a week or so ago, on one or two things `whatwg-url-without-unicode` might be able to handle more smoothly, with the removal of the `tr46` package? If that interests you, feel free to check it out, around [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/URL.20library/near/878195) and following.

**Comment by charpeni** [comment]
[`react-native-url-polyfill@1.2.0-rc.0`](https://www.npmjs.com/package/react-native-url-polyfill/v/1.2.0-rc.0) is now available with backward support for RN 0.60 & 0.59, and probably more!

Let me know how it goes for you so I can release it officially.

**Comment by chrisbobbe** [comment]
Confirmed that it worked for me on RN v0.59.10 and v0.60.6! Thank you!

**Comment by chrisbobbe** [comment]
A bit of housekeeping: I've renamed this issue "Start using proper URL objects..." so that the addition of `react-native-url-polyfill` (in a PR I'll file soon) will close this issue. Most of the discussion here was for settling on a strategy to use, and it'll be helpful to start fresh now that we've found a good one! 🙂 

Then, I'll open a new issue that can be resolved incrementally, where we track down and fix all our uses of string concatenation, regexes, etc., using this polyfill.

🎉
